File: 2626-count-the-number-of-good-subarrays/2626-count-the-number-of-good-subarrays.java
import java.util.HashMap;

class Solution {
    public long countGood(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int n = nums.length, left = 0;
        long ans = 0, pairCount = 0;

        for (int right = 0; right < n; right++) {
            int val = nums[right];
            int freq = map.getOrDefault(val, 0);
            pairCount += freq;
            map.put(val, freq + 1);

            while (pairCount >= k) {
                ans += (n - right);
                int leftVal = nums[left];
                int leftFreq = map.get(leftVal);
                map.put(leftVal, leftFreq - 1);
                pairCount -= leftFreq - 1;
                left++;
            }
        }

        return ans;
    }
}

File: 2626-count-the-number-of-good-subarrays/2626-count-the-number-of-good-subarrays.cpp
class Solution {
public:
    long long countGood(vector<int>& nums, int k) {
        unordered_map<int, int> mp;

        int n = nums.size(), right = 0, left = 0, pairCount = 0;
        long long ans = 0;

        for (int right = 0; right < n; right++) {
            pairCount += mp[nums[right]];
            mp[nums[right]]++;

            while (pairCount >= k) {
                ans += (n - right);
                mp[nums[left]]--;
                pairCount -= mp[nums[left]];
                left++;
            }
        }

        return ans;
    }
};